src/main/java/com/google/step/datamanager/DealManagerDatastore.java

```diff
@@ -134,40 +157,136 @@ public Deal updateDeal(Deal deal, List<String> tagNames) {
+  }
+
+  /** Retrieves deals posted by _ followed by user */
+  private List<Deal> getDealsPublishedByFollowedRestaurantsOrUsers(
+      List<Long> idsOfFollowedFieldName, String filterAttribute) {
+    List<Deal> dealResults = new ArrayList<>();
+    for (Long id : idsOfFollowedFieldName) {
+      Filter propertyFilter = new FilterPredicate(filterAttribute, FilterOperator.EQUAL, id);
+      Query query = new Query("Deal").setFilter(propertyFilter);
+      PreparedQuery pq = datastore.prepare(query);
+      for (Entity entity : pq.asIterable()) {
+        dealResults.add(readDeal(entity.getKey().getId()));
+      }
+    }
+    return dealResults;
   }
 
   /** Retrieves deals posted by users followed by user */
   @Override
   public List<Deal> getDealsPublishedByFollowedUsers(List<Long> userIds) {
-    return new ArrayList<Deal>();
+    return getDealsPublishedByFollowedRestaurantsOrUsers(userIds, "posterId");
   }
 
   /** Retrieves deals posted by restaurants followed by user */
   @Override
   public List<Deal> getDealsPublishedByFollowedRestaurants(List<Long> restaurantIds) {
-    return new ArrayList<Deal>();
+    return getDealsPublishedByFollowedRestaurantsOrUsers(restaurantIds, "restaurantId");
   }
 
   /** Retrieves deals posted by tags followed by user */
   @Override
   public List<Deal> getDealsPublishedByFollowedTags(List<Long> tagIds) {
-    return new ArrayList<Deal>();
+    List<Long> dealIdResults = new ArrayList<>();
+    for (Long id : tagIds) {
+      List<Long> dealIdsWithTag = dealTagManager.getDealIdsWithTag(id);
+      dealIdResults.addAll(dealIdsWithTag);
+    }
+    // Get rid of duplicate dealID (Deals with multiple tags)
+    List<Long> dealsWithoutDuplicates = new ArrayList<>(new HashSet<>(dealIdResults));
+    return readDeals(dealsWithoutDuplicates);
   }
 
+  /** Sorts deals based on votes (Highest to lowest) */
   @Override
   public List<Deal> sortDealsBasedOnVotes(List<Deal> deals) {
-    return new ArrayList<Deal>();
+    List<Map<String, Object>> dealWithVotesMaps = new ArrayList<Map<String, Object>>();
+    // Creates a list of maps with votes as an attribute to be sorted
+    for (Deal deal : deals) {
+      Map<String, Object> dealWithVotesMap = new HashMap<>();
+      dealWithVotesMap.put("votes", voteManager.getVotes(deal.id));
+      dealWithVotesMap.put("deal", deal);
+      dealWithVotesMaps.add(dealWithVotesMap);
+    }
+    return sortDealMapsBasedOnValue(dealWithVotesMaps, "votes");
   }
 
+  /** Method to sort a list of maps based on a value and return a list of deals */
+  private List<Deal> sortDealMapsBasedOnValue(
+      List<Map<String, Object>> dealMaps, String attribute) {
+    Collections.sort(
+        dealMaps,
+        new Comparator<Map<String, Object>>() {
+          @Override
+          public int compare(Map<String, Object> deal1, Map<String, Object> deal2) {
+            if (attribute.equals("hotScore")) // comparing hot score (double values)
```

Comment:
How about instead of sorting the Map<String,Object> with casting everywhere, maybe can create a new private class of pair? Like
```
class DealPair<T extends Comparable<T>> implements Comparable<DealPair<T>> {

  public final T key;
  public final Deal deal;

  public DealPair(T key, Deal deal) {
    this.key = key;
    this.deal = deal;
  }

  @Override
  public int compareTo(DealPair<T> other) {
    return key.compareTo(other.key);
  }
}
```

Reply:
ill do this for pr #67 cos i shifted over to the servlet